Fixes #32287 - setting index loads from registry

ezr-ondrej opened the pull request:
The values provided on the Setting index page are now served from
registry, with simple search API, that can be easily improved to support
more querying features.

ezr-ondrej committed:
Fixes #32287 - setting index loads from registry

The values provided on the Setting index page are now served from
registry, with simple search API, that can be easily improve to support
more querying features.

ezr-ondrej commented on app/helpers/settings_helper.rb:
```diff
@@ -1,39 +1,28 @@
-  end
-
   def grouped_settings(settings)
     settings.each_with_object({}) do |setting, memo|
```
please implement https://github.com/theforeman/foreman/pull/8002#discussion_r618333544

ezr-ondrej replied:
Note, this doesn't need the `compact` anymore as we are operating over the setting from registry.

ekohl commented on app/services/setting_registry.rb:
```diff
@@ -42,6 +57,15 @@ def []=(name, value)
 
+  # Returns all the categories used for settings
+  def categories
+    sticked_general = ActiveSupport::OrderedHash.new
```
https://api.rubyonrails.org/classes/ActiveSupport/OrderedHash.html mentions it's deprecated. I guess that's because Hash always behaves this way now? Quoting https://ruby-doc.org/core-2.5.0/Hash.html:
> Hashes enumerate their values in the order that the corresponding keys were inserted.

ezr-ondrej replied:
I've noticed the deprecation, but been wondering why that is... cool, in that case, simple hash is enough :+1:

lzap commented:
LGTM

lzap reviewed (changes requested):
The patch somehow breaks changing a value, I see an error and weird log:

![image](https://user-images.githubusercontent.com/49752/115827565-be0cba00-a40c-11eb-84ce-99556d210354.png)

```
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|app|2c71e1ba] Started PUT "/api/settings/proxy_request_timeout" for ::ffff:192.168.1.5 at 2021-04-23 08:19:53 +0200
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|app|2c71e1ba] Processing by Api::V2::SettingsController#update as JSON
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|app|2c71e1ba]   Parameters: {"setting"=>{"value"=>"[FILTERED]"}, "apiv"=>"v2", "id"=>"proxy_request_timeout"}
08:19:53 rails.1   | 2021-04-23T08:19:53 [D|app|2c71e1ba] Setting remote_addr has no definition, clean up your database
08:19:53 rails.1   | 2021-04-23T08:19:53 [D|app|2c71e1ba] Setting db_pending_migration has no definition, clean up your database
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|per|2c71e1ba] Current user set to admin (admin)
08:19:53 rails.1   | 2021-04-23T08:19:53 [D|tax|2c71e1ba] Current location set to Default Location
08:19:53 rails.1   | 2021-04-23T08:19:53 [D|tax|2c71e1ba] Current organization set to Default Organization
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|app|2c71e1ba]   Rendering api/v2/settings/update.json.rabl
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|app|2c71e1ba]   Rendered api/v2/settings/update.json.rabl (Duration: 0.9ms | Allocations: 742)
08:19:53 rails.1   | 2021-04-23T08:19:53 [D|app|2c71e1ba] Body: {"description":"Open and read timeout for HTTP requests from Foreman to Smart Proxy (in seconds)","category":"Setting::General","settings_type":"integer","default":60,"updated_at":"2021-04-23 08:19:43 +0200","id":"proxy_request_timeout","name":"proxy_request_timeout","full_name":"Smart Proxy request timeout","value":15,"category_name":"General","readonly":false,"config_file":null,"encrypted":false,"select_values":null}
08:19:53 rails.1   | 2021-04-23T08:19:53 [I|app|2c71e1ba] Completed 200 OK in 34ms (Views: 1.5ms | ActiveRecord: 2.6ms | Allocations: 24633)
```

ezr-ondrej commented:
> The patch somehow breaks changing a value, I see an error and weird log:

I've fixed this one, FTR we've changed the category names from 'Setting::General' to 'General', so I needed to change it in API update responses as well.

lzap commented on app/services/setting_registry.rb:
```diff
@@ -1,8 +1,23 @@
+  end
+
+  def search_for(query)
+    return self if query.blank?
```
Is returning self (SettingRegistry) really correct? I'd expect this would return a setting model rather than registry.

ezr-ondrej replied:
This returns self, that is (acts as) a collection, thus it mimics what `ActiveRecord.search_for()` does.

lzap commented on app/presenters/setting_presenter.rb:
```diff
@@ -52,10 +52,18 @@ def settings_type
+    else
+      description.include?(query) || name.include?(query) || full_name&.include?(query)
+    end
+  end
```
This will break compatibility, users do expect scoped_search. I have no problem with that, I think this is a minor problem. Just saying.

ezr-ondrej replied:
Indeed, that is a drawback to this solution.

lzap reviewed (commented):
Can you explain why are you doing this change?

Frankly, I don't know exactly what this change brings to the table. I do not know the Settings codebase well and I am confused, I am not sure to be the one to do the review. I can test this tho and if it works, well, this code can't be any worse I think.

lzap reviewed (approved):
Thank you. Merging.
